framework/src/variables/MooseLinearVariableFV.C

```diff
@@ -0,0 +1,261 @@
+  const auto & global_soln = (state.state == 0)
+                                 ? *this->_sys.currentSolution()
+                                 : this->_sys.solutionState(state.state, state.iteration_type);
+
+  return global_soln(elem_info.dofIndices()[this->_sys.number()][this->number()]);
+}
+
+template <typename OutputType>
+const VectorValue<Real> &
+MooseLinearVariableFV<OutputType>::gradSln(const ElemInfo & elem_info) const
+{
+  if (_needs_cell_gradients)
+  {
+    _cell_gradient.zero();
+    for (const auto i : make_range(this->_mesh.dimension()))
+      _cell_gradient(i) =
+          (*_grad_cache[i])(elem_info.dofIndices()[this->_sys.number()][this->number()]);
+  }
+
+  return _cell_gradient;
+}
+
+template <typename OutputType>
+VectorValue<Real>
+MooseLinearVariableFV<OutputType>::gradSln(const FaceInfo & fi, const StateArg & /*state*/) const
+{
+  const bool var_defined_on_elem = this->hasBlocks(fi.elem().subdomain_id());
+  const auto * const elem_one = var_defined_on_elem ? fi.elemInfo() : fi.neighborInfo();
+  const auto * const elem_two = var_defined_on_elem ? fi.neighborInfo() : fi.elemInfo();
+
+  const auto elem_one_grad = gradSln(*elem_one);
+
+  // If we have a neighbor then we interpolate between the two to the face.
+  if (elem_two && this->hasBlocks(elem_two->subdomain_id()))
+  {
+    const auto & elem_two_grad = gradSln(*elem_two);
+    return Moose::FV::linearInterpolation(elem_one_grad, elem_two_grad, fi, var_defined_on_elem);
+  }
+  else
+    return elem_one_grad;
+}
+
+template <typename OutputType>
+void
+MooseLinearVariableFV<OutputType>::initialSetup()
+{
+  MooseVariableField<OutputType>::initialSetup();
+
+  if (_needs_cell_gradients)
+  {
+    _grad_cache.clear();
+    for (const auto i : make_range(this->_mesh.dimension()))
+    {
+      libmesh_ignore(i);
+      _grad_cache.push_back(this->_sys.currentSolution()->zero_clone());
+    }
+  }
+
+  cacheBoundaryBCMap();
+}
+
+template <typename OutputType>
+typename MooseLinearVariableFV<OutputType>::ValueType
+MooseLinearVariableFV<OutputType>::evaluate(const FaceArg & face, const StateArg & state) const
+{
+  const FaceInfo * const fi = face.fi;
+
+  mooseAssert(fi, "The face information must be non-null");
+
+  const auto face_type = fi->faceType(std::make_pair(this->number(), this->sys().number()));
+  if (face_type == FaceInfo::VarFaceNeighbors::BOTH)
+    return Moose::FV::interpolate(*this, face, state);
+  else if (auto * bc_pointer = this->getBoundaryCondition(*fi->boundaryIDs().begin()))
+  {
+    mooseAssert(fi->boundaryIDs().size() == 1, "We should only have one boundary on every face.");
+
+    const auto * const original_face_info = bc_pointer->currentFaceInfo();
+    const auto original_face_type = bc_pointer->currentFaceType();
+
+    if (fi != original_face_info)
+      bc_pointer->setCurrentFaceInfo(fi, face_type);
+
+    const auto boundary_value = bc_pointer->computeBoundaryValue();
+
+    if (fi != original_face_info)
```

Comment:
why undo it?

Reply:
I just added it as a precaution for cases when somebody calls a function from a boundary condition which touches other boundaries and then still wants to evaluate some things on the same boundary. I removed it and everything works right now, so I'll leave it removed.